genfond/datalog_policy.py

```diff
@@ -0,0 +1,41 @@
+import logging
+import re
+
+log = logging.getLogger(__name__)
+
+
+class DatalogPolicyRule:
+
+    def __init__(self, head, tail):
+        match = re.search(r'^(\w+)\(([\w, ]+)\)$', head)
+        if not match:
+            raise ValueError(f'Invalid head: {head}')
+        self.name = match.groups()[0]
+        self.parameters = match.groups()[1].replace(' ', '').split(',')
+
+        self.tail_by_parameter = {name: [] for name in self.parameters}
+        for parameter, concept in tail or []:
+            if parameter not in self.parameters:
+                raise ValueError(f'Free variable {parameter} not in head parameters {self.parameters}!')
+            self.tail_by_parameter[parameter].append(concept.replace(' ', ''))
+
+    def __eq__(self, other):
+        return self.head == other.head and self.tail == other.tail #TODO
+
+    def __repr__(self):
+        tail = []
+        for parameter, concepts in self.tail_by_parameter.items():
+            tail.extend([str(concept).replace("0", parameter) for concept in concepts])
```

Comment:
This does not always work, e.g., if the concept is `c_oneof(Table)`, then the parameter can no longer be seen in the string representation, e.g.,
```
puton(X, Y, Z) :- c_one_of(Table), c_some(r_transitive_reflexive_closure(r_primitive(on,X,1)),c_projection(r_primitive(on_G,X,1),1)).
```

What about writing `X ∈ c_one_of(Table)` instead?

```suggestion
            tail.extend([f'{parameter ∈ {concept}' for concept in concepts])
```

Reply:
Did you test that? I get with this change:

```
tests/test_datalog_policy.py:1: in <module>
    from genfond.datalog_policy import DatalogPolicy, DatalogPolicyRule
E     File ".../genfond/genfond/datalog_policy.py", line 38
E       tail.extend([f'{parameter ∈ {concept}' for concept in concepts])
E                                 ^
E   SyntaxError: invalid character '∈' (U+2208)
```

Reply:
Ah ok nvm. Your suggestion is missing the closing `}`. I overlooked that